acceptance/tests/plan_ssh.rb

```diff
@@ -7,58 +8,95 @@
+    create_remote_file(bolt, "#{dir}/modules/test/tasks/hostname_nix.sh", <<-FILE)
+#!/bin/bash
+if [ ! -f /tmp/retry.txt ] && [ "$HOSTNAME" = '#{node}' ]; then
+  touch /tmp/retry.txt;
+  exit 1
+elif [ -f /tmp/retry.txt ]; then
+  rm /tmp/retry.txt
+fi
     FILE
+  end
+
+  step "create plan on bolt controller" do
     create_remote_file(bolt,
                        "#{dir}/modules/test/plans/my_unix_plan.pp", <<-FILE)
-    plan test::my_unix_plan($nodes) {
-      $nodes_array = $nodes.split(',')
-      notice("${run_task(test::a_unix, $nodes_array)}")
-      notice("${run_task(test::b_unix, $nodes_array)}")
-    }
+plan test::my_unix_plan($nodes) {
+  $array = $nodes.split(',')
+  $result = run_task(test::hostname_nix, $array,
+        'success_nodes' => [$array[0]],
+        '_catch_errors' =>true)
+  $retry = run_task(test::hostname_nix, $result.error_set.names,
+        'success_nodes' =>  $result.error_set.names)
+  return({ 'result' => $result, 'retry' => $retry })
+}
     FILE
   end
 
-  step "execute `bolt plan run` via SSH" do
-    user = ENV['SSH_USER']
-    password = ENV['SSH_PASSWORD']
-    nodes_csv = ssh_nodes.map(&:hostname).join(',')
+  step "execute `bolt plan run` via SSH with json output" do
     bolt_command = "bolt plan run test::my_unix_plan nodes=#{nodes_csv}"
-
     flags = {
       '-u'                     => user,
       '--modulepath'           => "#{dir}/modules",
       '-p'                     => password,
+      '--format'               => 'json',
       '--no-host-key-check'    => nil
     }
 
     result = bolt_command_on(bolt, bolt_command, flags)
-    message = "The plan was expected to notify but did not"
-    assert_match(/Notice:/, result.stdout, message)
-    ssh_nodes.each do |node|
-      message = "The plan was expected to mention the host #{node.hostname} with _output"
-      assert_match(/"node":"#{node.hostname}","status":"success"/, result.stdout, message)
-    end
+    assert_equal(0, result.exit_code,
+                 "The task did not exit with code 0 as expected")
 
-    ssh_nodes.each do |node|
-      # bolt plan return value unspecified
-      command = "cat /tmp/C100553_plan_artifact.txt"
-      on(node, command, accept_all_exit_codes: true) do |res|
-        cat_fail_msg = "The 'cat' command was not successful"
-        assert_equal(res.exit_code, 0, cat_fail_msg)
+    json = JSON.parse(result.stdout)
 
-        fail_msg = 'The observed contents of the plan artifact was unexpected'
-        assert_match(/Line one from task a/, res.stdout, fail_msg)
-        assert_match(/Line two from task b/, res.stdout, fail_msg)
-      end
+    # Verify that the first node failed on the first run
+    assert_equal(ssh_nodes[0].hostname, json['result'][0]['node'],
+                 "The hostname #{ssh_nodes[0].hostname} is not correct")
+    assert_equal('failure', json['result'][0]['status'],
+                 "The first node on the first run did not fail")
+
+    if ssh_nodes.length > 1
```

Comment:
Are we concerned that part of the test may not run in some configurations? Should we log a warning? Is there a standard for this?

Reply:
I haven't seen logging around different asserts being run for different configurations, but that seems like a good idea. Posited to some folks with more experience about standards, will let you know what I hear back.